document-store/src/main/java/org/hypertrace/core/documentstore/Collection.java

```diff
@@ -157,13 +159,21 @@ CloseableIterator<Document> bulkUpsertAndReturnOlderDocuments(Map<Key, Document>
   void drop();
 
   /**
-   * create a new document if one doesn't exists with the key
+   * Create a new document if one doesn't violate any unique constraints (including the unique key
+   * constraint)
    *
    * @param key Unique key of the document in the collection.
    * @param document Document to be created.
    * @return an instance of {@link CreateResult}
+   * @throws DuplicateDocumentException if either
+   *     <ul>
+   *       <li>a document with the given key already exists or
+   *       <li>the given document violates an unique constraint
+   *     </ul>
+   *
+   * @throws IOException if there was any other exception while creating the document
    */
-  CreateResult create(Key key, Document document) throws IOException;
+  CreateResult create(Key key, Document document) throws DuplicateDocumentException, IOException;
```

Comment:
> Did we consider enhancing it with failure code, and reason?

No use-case for failure codes as of now. Even if we want failure codes, we still want a custom exception since IOException internally doesn't have a way to define codes.

> Do we want to take different client actions based on different exceptions?

Yes. In fact, the existing implementation is buggy. It never returns a `CreateResult` with a false value. We could fix that, but at the cost of breaking backward compatibility. That's the reason it was decided to use a custom exception class.